src/durmod/kaldi-durmod.cc

```diff
@@ -0,0 +1,463 @@
+        binary_feats = binary_feats_.find(phone)->second;
+    }
+
+    // phone-id feature
+    feat_elements.push_back(std::make_pair(feat_idx + phone_id, 1.0));
+    feat_idx += num_phone_identities_;
+
+    // binary features (from extra questions for eg.)
+    for (int j = 0; j < binary_feats.size(); j++) {
+      feat_elements.push_back(std::make_pair(feat_idx + binary_feats[j], 1.0));
+    }
+    feat_idx += num_binary_features_;
+
+    // duration features
+    if (i < phone_index) {
+      feat_elements.push_back(
+                             std::make_pair(feat_idx++,
+                                            NormalizeDuration(phone_duration)));
+    }
+  }
+  SparseVector<BaseFloat> tmp(feature_dim_, feat_elements);
+  feat->CopyFromSvec(tmp);
+}
+
+void PhoneDurationEgsMaker::MakeNnetExample(
+                    const std::vector<std::pair<int32, int32> > &phone_context,
+                    int phone_index,
+                    NnetExample *eg) const {
+  SparseVector<BaseFloat> feat;
+  MakeFeatureVector(phone_context, phone_index, &feat);
+  int32 phone_duration = phone_context[phone_index].second;
+  SparseMatrix<BaseFloat> feat_mat(1, feat.Dim());
+  feat_mat.SetRow(0, feat);
+  int32 output_dim = max_duration_;
+  Posterior output_elements(1);
+
+  // The nodes at the output of the network are indexed from
+  // 0 to (max_duration_ - 1). Index 0 is for duration = 1.
+  int32 duration_id = (phone_duration > max_duration_) ?
+                                                       (max_duration_ - 1):
+                                                       (phone_duration - 1);
+  output_elements[0].push_back(std::make_pair(duration_id, 1.0));
+  SparseMatrix<BaseFloat> output_mat(output_dim, output_elements);
+
+  NnetIo input, output;
+  input.name = "input";
+  input.features = feat_mat;
+  input.indexes.resize(1);
+  output.name = "output";
+  output.features = output_mat;
+  output.indexes.resize(1);
+  eg->io.push_back(input);
+  eg->io.push_back(output);
+}
+
+void PhoneDurationEgsMaker::AlignmentToNnetExamples(
+                      const std::vector<std::pair<int32, int32> > &alignment,
+                      std::vector<NnetExample> *egs) const {
+// These lines are commented because we need edge examples too:
+//  if (alignment.size() < (model.left_ctx + model.right_ctx + 1)) {
+//    return;
+//  }
+  for (int i = 0; i < alignment.size(); i++) {
+    NnetExample eg;
+    MakeNnetExample(alignment, i, &eg);
+    // eg.io[0].indexes[0].t = i;
+    // eg.io[1].indexes[0].t = i;
+    egs->push_back(eg);
+  }
+}
+
+void PhoneDurationEgsMaker::InitFeatureMaker(const PhoneDurationModel &model) {
+  num_binary_features_ = model.questions_.size();
+  num_phone_identities_ = model.roots_.size() + 1;  // id=0 is for not-available
+                                                    // phones (i.e. null phones
+                                                    // which occur at edges).
+  left_ctx_ = model.left_context_;
```

Comment:
please don't use this spelling of context: just left_context_.  the google style guide discourages abbreviations like this.